ASPL-836: ASE Postman Collection

vishalhcl-5960 opened the pull request:
Implementation of the AppScan Enterprise Postman collection function.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanEnterpriseBuildStep.java:
```diff
@@ -788,6 +974,50 @@ public FormValidation doCheckJobName(@QueryParameter String jobName) {
+		}
+
+		public FormValidation doCheckPostmanCollectionFile(@QueryParameter String postmanCollectionFile) {
+			if(postmanCollectionFile != null && !postmanCollectionFile.isEmpty()) {
```
Rather than a nested "if" block, this should just be logically "anded" with the other conditions.

vishalhcl-5960 replied:
Okay, sure.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanEnterpriseBuildStep.java:
```diff
@@ -788,6 +974,50 @@ public FormValidation doCheckJobName(@QueryParameter String jobName) {
+		}
+
+		public FormValidation doCheckEnvironmentalVariablesFile(@QueryParameter String environmentalVariablesFile) {
+			if(environmentalVariablesFile != null && !environmentalVariablesFile.isEmpty()) {
```
Rather than a nested "if" block, this should just be logically "anded" with the other conditions.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanEnterpriseBuildStep.java:
```diff
@@ -788,6 +974,50 @@ public FormValidation doCheckJobName(@QueryParameter String jobName) {
+		}
+
+		public FormValidation doCheckGlobalVariablesFile(@QueryParameter String globalVariablesFile) {
+			if(globalVariablesFile != null && !globalVariablesFile.isEmpty()) {
```
Rather than a nested "if" block, this should just be logically "anded" with the other conditions.

vishalhcl-5960 replied:
Okay, sure.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanEnterpriseBuildStep.java:
```diff
@@ -788,6 +974,50 @@ public FormValidation doCheckJobName(@QueryParameter String jobName) {
+		}
+
+		public FormValidation doCheckAdditionalFiles(@QueryParameter String additionalFiles) {
+			if(additionalFiles != null && !additionalFiles.isEmpty()) {
```
Rather than a nested "if" block, this should just be logically "anded" with the other conditions.

mattmurp reviewed (changes requested):
The number of conditionals in this class and the amount of logic is very concerning and these latest changes are adding more of the same. This appears due to the fact that there are 4 different scans types, all of which require their own specific set of properties. We should refactor this to have an IScanType interface and 4 concrete implementations of that interface. The AppScanEnterpriseBuildStep class should not be responsible for all of this logic. We should also use an enum to represent the scan type, as opposed to using string representations of integers, which is very difficult to understand and maintain.

vishalhcl-5960 commented:
> The number of conditionals in this class and the amount of logic is very concerning and these latest changes are adding more of the same. This appears due to the fact that there are 4 different scans types, all of which require their own specific set of properties. We should refactor this to have an IScanType interface and 4 concrete implementations of that interface. The AppScanEnterpriseBuildStep class should not be responsible for all of this logic. We should also use an enum to represent the scan type, as opposed to using string representations of integers, which is very difficult to understand and maintain.

Implemented the suggested changes by creating three dedicated classes for each scan type, mirroring the structure of the corresponding scanner classes used in AppScanBuildStep. Additionally, introduced a readResolve() method to ensure backward compatibility with previously serialized configurations.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/ScanModeFactory.java:
```diff
@@ -0,0 +1,17 @@
+
+public class ScanModeFactory {
+
+    public static ScanMode getScanTypeImplementation(String scanType, String loginType, String userName, String password, String trafficFile, String exploreData) {
```
Not a must, but this method would be better if it were called "getScanMode", since this is a ScanModeFactory. It's not clear what is meant by "getScanTypeImplementation".

vishalhcl-5960 replied:
Okay, I will rename it to "getScanMode".

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/ScanModeFactory.java:
```diff
@@ -0,0 +1,17 @@
+public class ScanModeFactory {
+
+    public static ScanMode getScanTypeImplementation(String scanType, String loginType, String userName, String password, String trafficFile, String exploreData) {
+        switch (scanType) {
```
Is there a reason we're using String representations of numbers here? Can we use an enum instead?

vishalhcl-5960 replied:
Sure, now I am using enums for the comparison.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/TestOnly.java:
```diff
@@ -0,0 +1,118 @@
+
+        @Override
+        public String getDisplayName() {
+            return "Test Only";
```
Can you use the TEST_ONLY constant defined above here?

vishalhcl-5960 replied:
Okay, sure.

mattmurp commented on src/main/resources/com/hcl/appscan/jenkins/plugin/Messages.properties:
```diff
@@ -56,7 +56,8 @@ error.invalid.scan.id.rescan.allowed.ui = Rescan disabled by admin policy.
 error.token.authentication = Could not validate your Key ID and Secret. Please check and try again.
 error.scan.status.unstable= One scan execution failed. Refer to the build summary page for the results of the successful scan.
-
+error.file.type.invalid.json= Invalid file type. Only .json file is supported.
```
The 2 message strings should be combined into a single message that takes the type as a paramter. For example:
`error.file.type.invalid=Invalid file type. Only {0} files are suported.`

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/PostmanCollection.java:
```diff
@@ -0,0 +1,146 @@
+            return "Postman Collection";
+        }
+
+        public FormValidation doCheckPostmanCollectionFile(@QueryParameter String postmanCollectionFile) {
```
The 4 "doCheckXXXFile" all have very similar logic. I suggest moving that logic into a private method that all 4 of these methods call, rather than duplicating the logic 4 times.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/PostmanCollection.java:
```diff
@@ -0,0 +1,146 @@
+
+
+    @Override
+    public Map<String, String> configureScanProperties(Map<String, String> properties, VariableResolver<String> resolver) {
```
The 5 conditionals in this method all do essentially the same thing, just working on different member variables. This logic should be pulled out into a private function that can be called 5 times, rather than duplicating the code 5 times.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/FullScan.java:
```diff
@@ -0,0 +1,109 @@
+        return m_exploreData;
+    }
+
+    public String isLoginType(String loginTypeName) {
```
Is there a reason isLoginType returns a String instead of a boolean? If it needs to be a String the logic could be cleaned up to just be:
`return loginTypeName.equalsIgnoreCase(m_loginType) || loginTypeName.equals("Manual") ? "true" : "";`

vishalhcl-5960 replied:
Yes, LoginType needs to be a String. Updating the methods in both classes as per the suggestion.

mattmurp reviewed (changes requested):
In addition to the comments inline, there are a number of hardcoded strings duplicated throughout these files that should be made constants ("scanType", "loginType", "Manual", etc.).

vishalhcl-5960 commented:
> In addition to the comments inline, there are a number of hardcoded strings duplicated throughout these files that should be made constants ("scanType", "loginType", "Manual", etc.).

Sure, created constants for the frequently used strings.

mattmurp commented on src/main/java/com/hcl/appscan/jenkins/plugin/builders/AppScanEnterpriseBuildStep.java:
```diff
@@ -545,7 +527,7 @@ public IResultsProvider call() throws AbortException {
 						while(m_scanStatus != null && (m_scanStatus.equalsIgnoreCase("Waiting to Auto Run") || m_scanStatus.equalsIgnoreCase("Waiting to Run")
 								|| m_scanStatus.equalsIgnoreCase("Starting") || m_scanStatus.equalsIgnoreCase("Running")
-								|| m_scanStatus.equals("Post Processing") || m_scanStatus.equals("Waiting to Generate Results") || m_scanStatus.equals("Generating Results"))) {
+								|| m_scanStatus.equals("Post Processing") || m_scanStatus.equals("Waiting to Generate Results") || m_scanStatus.equals("Generating Results") || m_scanStatus.equalsIgnoreCase("Canceling"))) {
```
Given that there are now 8 different statuses being checked for to see if the scan is still executing, is there a smaller set that we could check for to indicate that the scan is completed?

vishalhcl-5960 replied:
Sure, replacing the check with these conditions: "!(m_scanStatus.equalsIgnoreCase(CoreConstants.READY) || m_scanStatus.equalsIgnoreCase(CoreConstants.FAILED) || m_scanStatus.startsWith(CoreConstants.SUSPENDED)))"

mvjainhcl reviewed (commented):
Test

mvjainhcl commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/FullScan.java:
```diff
@@ -0,0 +1,110 @@
+    public Map<String, String> configureScanProperties(Map<String, String> properties, VariableResolver<String> resolver) {
+        properties.put(ScanModeConstants.SCAN_TYPE, FULL_SCAN);
+        properties.put(ScanModeConstants.LOGIN_TYPE, m_loginType);
+        properties.put("exploreData", resolver == null || m_exploreData.equals(EMPTY) ? m_exploreData : resolvePath(m_exploreData, resolver));
```
this line can throw a NullPointerException if m_exploreData is null.

vishalhcl-5960 replied:
The potential for a NullPointerException exists if m_exploreData is null and .equals() is invoked on it. I've now updated the logic to avoid that by ensuring the EMPTY.equals(m_exploreData) check is safe. Additionally, we already perform a null check for m_exploreData in the updateJob() method of the SDK before the API call is made, so this scenario should already be guarded at a higher level as well.

mvjainhcl commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/FullScan.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @DataBoundSetter
+    public void setAccessId(String userName) {
```
The field m_userName is set via setAccessId and getAccessId. Would be good to do renaming for clarity and consistency (either userName or accessId)

mvjainhcl commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/PostmanCollection.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private void addResolvedProperty(Map<String, String> properties, String key, String value, VariableResolver resolver) {
+        if (!isNullOrEmpty(value)) return;
```
Please check if there is any logic error here. It will simply return if the value is not empty or null. Shouldn't the logic be - if (isNullOrEmpty(value)) return; to skip adding empty values?

mvjainhcl commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/PostmanCollection.java:
```diff
@@ -0,0 +1,147 @@
+        this("", "", "", "", "");
+    }
+
+    public PostmanCollection(String postmanCollectionFile, String additionalDomains, String environmentalVariablesFile, String globalVariablesFile, String additionaFiles) {
```
The parameter "additionaFiles" is misspelled; it should be additionalFiles for consistency.

vishalhcl-5960 commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/PostmanCollection.java:
```diff
@@ -0,0 +1,147 @@
+    @Override
+    public Map<String, String> configureScanProperties(Map<String, String> properties, VariableResolver<String> resolver) {
+        properties.put(ScanModeConstants.SCAN_TYPE, POSTMAN_COLLECTION);
+        addResolvedProperty(properties, "postmanCollectionFile", m_postmanCollectionFile, resolver);
```
Okay, sure. Added them as constants in the ScanModeConstants class.

vishalhcl-5960 commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/FullScan.java:
```diff
@@ -0,0 +1,110 @@
+        properties.put(ScanModeConstants.LOGIN_TYPE, m_loginType);
+        properties.put("exploreData", resolver == null || m_exploreData.equals(EMPTY) ? m_exploreData : resolvePath(m_exploreData, resolver));
+        if (ScanModeConstants.MANUAL.equals(m_loginType)) {
+            properties.put("trafficFile", resolver == null || m_trafficFile.equals(EMPTY) ? m_trafficFile : resolvePath(m_trafficFile, resolver));
```
Okay, sure. Added them as constants in the ScanModeConstants class.

vishalhcl-5960 commented on src/main/java/com/hcl/appscan/jenkins/plugin/scanModes/ScanMode.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public abstract class ScanMode extends AbstractDescribableImpl<ScanMode> implements ScannerConstants, Serializable {
```
Okay, sure.
